src/storage.rs

```diff
@@ -0,0 +1,195 @@
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use serde_json::{from_str, to_string};
+use std::fs::{File, OpenOptions};
+use std::io::{Read, Write, Seek, SeekFrom};
+use std::ops::{Deref, DerefMut};
+use std::fmt;
+
+/// TODO: Write some docs
+#[derive(Debug)]
+pub struct StorageReadGuard<'a, T: Serialize + DeserializeOwned + 'a> {
+    cache: &'a Storage<T>,
+}
+
+impl<'a, T: Serialize + DeserializeOwned> StorageReadGuard<'a, T> {
+    fn new(cache: &'a Storage<T>) -> Self {
+        Self { cache }
+    }
+}
+
+impl<'a, T: Serialize + DeserializeOwned + 'a> Deref for StorageReadGuard<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &self.cache.data
+    }
+}
+
+impl<'a, T: 'a + Serialize + DeserializeOwned + fmt::Display> fmt::Display for StorageReadGuard<'a, T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        (**self).fmt(f)
+    }
+}
+
+/// TODO: Write some docs
```

Comment:
Remove todo comment, create story instead

Reply:
Buried this at the end of the PR description, but I was planning on filling out the doc TODOs before this would get merged, I wanted to get feedback about the general approach first though